File: sidecar/vistructum_ml/features.py
import numpy as np

from .contract import FULLSCAN, GRID, HEIGHT_CLIP, HEIGHT_CONTEXT, HEIGHT_SAMPLE_STEP, MASK, STRIDE
from .scene import UNKNOWN

HEIGHT_ZERO = HEIGHT_CLIP
LUMINANCE_PAD = 128


def mask_features(scene):
    return scene.modified.astype(np.uint8)[None]


def relative_height(scene):
    known = scene.blocks != UNKNOWN
    heights = np.where(known, scene.heights, 0).astype(np.float32)
    reach = HEIGHT_CONTEXT // 2
    padded = np.pad(heights, reach, mode="edge")
    padded_known = np.pad(known, reach, mode="edge")
    offsets = range(-reach, reach + 1, HEIGHT_SAMPLE_STEP)
    rows, cols = heights.shape
    samples = []
    for dr in offsets:
        for dc in offsets:
            window = padded[reach + dr:reach + dr + rows, reach + dc:reach + dc + cols]
            valid = padded_known[reach + dr:reach + dr + rows, reach + dc:reach + dc + cols]
            samples.append(np.where(valid, window, np.nan))
    ground = np.nanmedian(np.stack(samples), axis=0) if samples else heights
    ground = np.where(np.isnan(ground), heights, ground)
    relative = np.clip(np.rint(heights - ground), -HEIGHT_CLIP, HEIGHT_CLIP)
    return np.where(known, relative + HEIGHT_ZERO, HEIGHT_ZERO).astype(np.uint8)


def block_boundaries(scene):
    blocks = scene.blocks
    edge = np.zeros(blocks.shape, dtype=bool)
    vertical = blocks[1:, :] != blocks[:-1, :]
    horizontal = blocks[:, 1:] != blocks[:, :-1]
    edge[1:, :] |= vertical
    edge[:-1, :] |= vertical
    edge[:, 1:] |= horizontal
    edge[:, :-1] |= horizontal
    edge &= blocks != UNKNOWN
    return edge.astype(np.uint8)


def height_steps(scene):
    """1 on both sides of every step of at least one block between 4-neighbours: the outline of anything raised or
    sunk, even when it is built from the ground's own block and so has no block-id boundary and no luminance change"""
    known = scene.blocks != UNKNOWN
    heights = scene.heights
    edge = np.zeros(heights.shape, dtype=bool)
    vertical = (heights[1:, :] != heights[:-1, :]) & known[1:, :] & known[:-1, :]
    horizontal = (heights[:, 1:] != heights[:, :-1]) & known[:, 1:] & known[:, :-1]
    edge[1:, :] |= vertical
    edge[:-1, :] |= vertical
    edge[:, 1:] |= horizontal
    edge[:, :-1] |= horizontal
    return edge.astype(np.uint8)


def fullscan_features(scene):
    luminance = np.where(scene.blocks != UNKNOWN, scene.luminance, LUMINANCE_PAD).astype(np.uint8)
    return np.stack([relative_height(scene), block_boundaries(scene), luminance, height_steps(scene)])


EXTRACTORS = {MASK.name: mask_features, FULLSCAN.name: fullscan_features}
PAD_VALUES = {MASK.name: (0,), FULLSCAN.name: (HEIGHT_ZERO, 0, LUMINANCE_PAD, 0)}


def extract(kind, scene):
    return EXTRACTORS[kind](scene)


def pad_to_grid(kind, features):
    channels, rows, cols = features.shape
    out_rows = max(rows, GRID)
    out_cols = max(cols, GRID)
    if (out_rows, out_cols) == (rows, cols):
        return features
    out = np.empty((channels, out_rows, out_cols), dtype=np.uint8)
    for channel, value in enumerate(PAD_VALUES[kind]):
        out[channel] = value
    out[:, :rows, :cols] = features
    return out


def window_origins(length, size=GRID, stride=STRIDE):
    if length <= size:
        return [0]
    origins = list(range(0, length - size + 1, stride))
    if origins[-1] != length - size:
        origins.append(length - size)
    return origins


def windows(kind, features, size=GRID, stride=STRIDE):
    padded = pad_to_grid(kind, features)
    _, rows, cols = padded.shape
    positions = [(top, left) for top in window_origins(rows, size, stride)
                 for left in window_origins(cols, size, stride)]
    stack = np.stack([padded[:, top:top + size, left:left + size] for top, left in positions])
    return positions, stack

File: vistructum-inference/src/main/java/de/kylekreuter/vistructum/inference/Features.java
package de.kylekreuter.vistructum.inference;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public final class Features {

    private static final int HEIGHT_ZERO = Contract.HEIGHT_CLIP;

    private Features() {
    }

    public static FeatureMap extract(ModelKind kind, SurfaceScene scene) {
        return switch (kind) {
            case MASK -> new FeatureMap(1, scene.height(), scene.width(), scene.modified().clone());
            case FULLSCAN -> fullscan(scene);
        };
    }

    public static WindowBatch windows(ModelKind kind, FeatureMap features) {
        FeatureMap padded = padToGrid(kind, features);
        List<Integer> rowOrigins = origins(padded.rows());
        List<Integer> colOrigins = origins(padded.cols());
        int count = rowOrigins.size() * colOrigins.size();
        int grid = Contract.GRID;
        int[] tops = new int[count];
        int[] lefts = new int[count];
        byte[] values = new byte[count * padded.channels() * grid * grid];
        int window = 0;
        for (int top : rowOrigins) {
            for (int left : colOrigins) {
                tops[window] = top;
                lefts[window] = left;
                for (int channel = 0; channel < padded.channels(); channel++) {
                    for (int row = 0; row < grid; row++) {
                        int source = (channel * padded.rows() + top + row) * padded.cols() + left;
                        int target = ((window * padded.channels() + channel) * grid + row) * grid;
                        System.arraycopy(padded.values(), source, values, target, grid);
                    }
                }
                window++;
            }
        }
        return new WindowBatch(padded.channels(), tops, lefts, values);
    }

    static List<Integer> origins(int length) {
        int grid = Contract.GRID;
        List<Integer> origins = new ArrayList<>();
        if (length <= grid) {
            origins.add(0);
            return origins;
        }
        for (int origin = 0; origin <= length - grid; origin += Contract.STRIDE) {
            origins.add(origin);
        }
        if (origins.getLast() != length - grid) {
            origins.add(length - grid);
        }
        return origins;
    }

    private static FeatureMap padToGrid(ModelKind kind, FeatureMap features) {
        int rows = Math.max(features.rows(), Contract.GRID);
        int cols = Math.max(features.cols(), Contract.GRID);
        if (rows == features.rows() && cols == features.cols()) {
            return features;
        }
        byte[] values = new byte[features.channels() * rows * cols];
        for (int channel = 0; channel < features.channels(); channel++) {
            Arrays.fill(values, channel * rows * cols, (channel + 1) * rows * cols, (byte) kind.padValue(channel));
            for (int row = 0; row < features.rows(); row++) {
                System.arraycopy(features.values(), (channel * features.rows() + row) * features.cols(), values,
                        (channel * rows + row) * cols, features.cols());
            }
        }
        return new FeatureMap(features.channels(), rows, cols, values);
    }

    private static FeatureMap fullscan(SurfaceScene scene) {
        int count = scene.width() * scene.height();
        byte[] values = new byte[4 * count];
        byte[] relative = relativeHeight(scene);
        byte[] boundaries = blockBoundaries(scene);
        byte[] steps = heightSteps(scene);
        for (int i = 0; i < count; i++) {
            values[i] = relative[i];
            values[count + i] = boundaries[i];
            values[2 * count + i] = known(scene, i) ? scene.luminance()[i] : (byte) Contract.LUMINANCE_PAD;
            values[3 * count + i] = steps[i];
        }
        return new FeatureMap(4, scene.height(), scene.width(), values);
    }

    private static boolean known(SurfaceScene scene, int index) {
        return scene.blocks()[index] != SurfaceScene.UNKNOWN;
    }

    static byte[] relativeHeight(SurfaceScene scene) {
        int rows = scene.height();
        int cols = scene.width();
        int reach = Contract.HEIGHT_CONTEXT / 2;
        byte[] out = new byte[rows * cols];
        double[] samples = new double[(2 * reach / Contract.HEIGHT_SAMPLE_STEP + 1)
                * (2 * reach / Contract.HEIGHT_SAMPLE_STEP + 1)];
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                int index = scene.index(row, col);
                if (!known(scene, index)) {
                    out[index] = (byte) HEIGHT_ZERO;
                    continue;
                }
                int valid = 0;
                for (int dr = -reach; dr <= reach; dr += Contract.HEIGHT_SAMPLE_STEP) {
                    for (int dc = -reach; dc <= reach; dc += Contract.HEIGHT_SAMPLE_STEP) {
                        int sample = scene.index(clamp(row + dr, rows), clamp(col + dc, cols));
                        if (known(scene, sample)) {
                            samples[valid++] = scene.heights()[sample];
                        }
                    }
                }
                double height = scene.heights()[index];
                double ground = valid == 0 ? height : median(samples, valid);
                double relative = Math.max(-Contract.HEIGHT_CLIP, Math.min(Contract.HEIGHT_CLIP, Math.rint(height - ground)));
                out[index] = (byte) (relative + HEIGHT_ZERO);
            }
        }
        return out;
    }

    private static int clamp(int value, int length) {
        return Math.max(0, Math.min(length - 1, value));
    }

    private static double median(double[] samples, int count) {
        Arrays.sort(samples, 0, count);
        int middle = count / 2;
        return count % 2 == 1 ? samples[middle] : (samples[middle - 1] + samples[middle]) / 2.0;
    }

    static byte[] blockBoundaries(SurfaceScene scene) {
        int rows = scene.height();
        int cols = scene.width();
        short[] blocks = scene.blocks();
        boolean[] edge = new boolean[rows * cols];
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                int index = scene.index(row, col);
                if (row + 1 < rows && blocks[index] != blocks[index + cols]) {
                    edge[index] = true;
                    edge[index + cols] = true;
                }
                if (col + 1 < cols && blocks[index] != blocks[index + 1]) {
                    edge[index] = true;
                    edge[index + 1] = true;
                }
            }
        }
        byte[] out = new byte[rows * cols];
        for (int i = 0; i < out.length; i++) {
            out[i] = (byte) (edge[i] && known(scene, i) ? 1 : 0);
        }
        return out;
    }

    static byte[] heightSteps(SurfaceScene scene) {
        int rows = scene.height();
        int cols = scene.width();
        short[] heights = scene.heights();
        byte[] out = new byte[rows * cols];
        for (int row = 0; row < rows; row++) {
            for (int col = 0; col < cols; col++) {
                int index = scene.index(row, col);
                if (row + 1 < rows && heights[index] != heights[index + cols] && known(scene, index)
                        && known(scene, index + cols)) {
                    out[index] = 1;
                    out[index + cols] = 1;
                }
                if (col + 1 < cols && heights[index] != heights[index + 1] && known(scene, index)
                        && known(scene, index + 1)) {
                    out[index] = 1;
                    out[index + 1] = 1;
                }
            }
        }
        return out;
    }
}
